xc/xenops_server_xen.ml

```diff
@@ -916,6 +902,35 @@ module Featureset = struct
     |> String.concat ":"
 end
 
+let debug_featuresets xc name featureset default featureset_max index_max =
+  (* For debugging purposes retrieve the default featureset and compare it *)
+  debug "%s Max featureset = %a" name Featureset.pp featureset_max ;
+  let featureset_default = Xenctrl.get_cpu_featureset xc default in
+  debug "%s Default featureset = %a" name Featureset.pp featureset_default ;
+  debug "%s Dynamic set = %a" name Featureset.pp featureset ;
+
+  (* sanity checks: our featureset shouldn't be more than max! *)
+  let ours_minus_max = Featureset.diff featureset featureset_max in
+  if not @@ Array.for_all (fun x -> Int64.equal x 0L) ours_minus_max then (
+    debug "%s (DynamicSet - Max) = %a" name Featureset.pp ours_minus_max ;
+    warn
+      "Our default policy has CPUID features that are not present in Max \
+       policy! (see above)"
+  ) ;
+  debug "%s (Max-Default) = %a" name Featureset.pp
+    (Featureset.diff featureset_max featureset_default)
+
+let get_policy_featureset xc name featureset default index_max =
+  (* The migration-safety policy of max CPUID we can accept *)
+  let featureset_max = Xenctrl.get_cpu_featureset xc index_max in
+
+  debug_featuresets xc name featureset default featureset_max index_max ;
+
+  (* just in case OR the max into our featureset instead of taking max as is *)
+  let migration_safe = Featureset.map2 Int64.logor featureset featureset_max in
```

Comment:
I did it this way because otherwise the newly computed max would have less bits in than we had in the default (my sanity test above triggered saying we have bits set in default missing from max), although this might be fixed once the 4.7 backward compat bits get moved, I'll test again on a host. You are right, those are probably bugs we should fix in the computation of Max in that case.